File: Leet-Code/minDepthOfTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <queue>
#include <tuple>

class Solution {
public:
    int minDepth(TreeNode* root) {
        if(NULL == root)
            return 0;

        queue<tuple<TreeNode*, int> > q;
        q.push(tuple<TreeNode*, int>(root, 1));
        
        tuple<TreeNode*, int> node;
        TreeNode* child; 

        while(!q.empty())
        {
            node = q.front();
            child = get<0>(node);

            if(child->left == NULL && child->right == NULL)
            {
                return get<1>(node);
            }
            if(child->left != NULL)
            {
                q.push(tuple<TreeNode*, int>(child->left, get<1>(node) + 1));
            }
            if(child->right != NULL)
            {
                q.push(tuple<TreeNode*, int>(child->right, get<1>(node) + 1));
            }

            q.pop();
        }
        return 0;
    }
};

File: Leet-Code/minDepthOfTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from queue import Queue

class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0

        q = Queue(0)
        q.put((root, 1))

        while True:
            node, depth = q.get()

            if node.left is None and node.right is None:
                break

            if node.left is not None:
                q.put((node.left, depth + 1))
            if node.right is not None:
                q.put((node.right, depth + 1))
        
        return depth